File: hash/js/rosetta-image-hash/src/decodePng.ts
import { PNG } from "pngjs";

import type { RgbImage } from "./hash.js";

/**
 * Decode PNG bytes into an RgbImage with channels=4 (RGBA from pngjs).
 * Internal algorithms composite alpha against opaque black per PIL convert('RGB').
 *
 * Node-friendly path: pass a Uint8Array or Node Buffer; we wrap appropriately.
 * pngjs ships only a Node-flavored sync API, so this helper is Node-only.
 * Browser callers should use `canvas.getImageData(...)` and construct RgbImage themselves.
 */
export function decodePng(bytes: Uint8Array): RgbImage {
	// PNG.sync.read needs a Node Buffer; wrap the Uint8Array
	const buf = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);
	const png = PNG.sync.read(buf);
	// png.data is a Buffer (Uint8Array view) of length width*height*4 (RGBA)
	return {
		width: png.width,
		height: png.height,
		data: new Uint8Array(png.data),
		channels: 4,
	};
}

File: hash/swift/RosettaImageHash/Sources/RosettaImageHash/DecodePNG.swift
import Foundation
import PNG

/// Decode PNG bytes into an RGBImage with channels = .rgba.
/// Internal algorithms composite alpha against opaque black per PIL `convert('RGB')`.
public func decodePNG(_ bytes: [UInt8]) throws -> RGBImage {
	var blob = PNGBlob(data: bytes)
	let image: PNG.Image = try .decompress(stream: &blob)
	let rgba: [PNG.RGBA<UInt8>] = image.unpack(as: PNG.RGBA<UInt8>.self)
	let width = image.size.x
	let height = image.size.y

	// Flatten [PNG.RGBA<UInt8>] -> [UInt8] in RGBA order
	var data = [UInt8](repeating: 0, count: rgba.count * 4)
	for i in 0..<rgba.count {
		data[i * 4]     = rgba[i].r
		data[i * 4 + 1] = rgba[i].g
		data[i * 4 + 2] = rgba[i].b
		data[i * 4 + 3] = rgba[i].a
	}
	return RGBImage(width: width, height: height, data: data, channels: .rgba)
}

/// In-memory byte stream conforming to swift-png's bytestream protocol.
internal struct PNGBlob: PNG.BytestreamSource {
	let data: [UInt8]
	var position: Int

	init(data: [UInt8]) {
		self.data = data
		self.position = 0
	}

	mutating func read(count: Int) -> [UInt8]? {
		guard position + count <= data.count else { return nil }
		defer { position += count }
		return Array(data[position..<(position + count)])
	}
}
